File: extensions/loyalty-discount-function/src/cart_lines_discounts_generate_run.rs
use super::schema;
use shopify_function::Result;
use shopify_function::prelude::*;

#[shopify_function]
fn cart_lines_discounts_generate_run(
    input: schema::cart_lines_discounts_generate_run::Input,
) -> Result<schema::CartLinesDiscountsGenerateRunResult> {
    let redeemed_points = input
        .cart()
        .attribute()
        .and_then(|attribute| attribute.value().as_ref())
        .map(|value| value.trim())
        .filter(|value| !value.is_empty())
        .and_then(|value| value.parse::<i64>().ok())
        .filter(|points| *points > 0);

    if redeemed_points.is_none() {
        return Ok(schema::CartLinesDiscountsGenerateRunResult { operations: vec![] });
    }

    // Placeholder: discount calculation and operations will be added next.
    Ok(schema::CartLinesDiscountsGenerateRunResult { operations: vec![] })
}

File: extensions/loyalty-discount-function/src/cart_lines_discounts_generate_run.js
import {
  DiscountClass,
  OrderDiscountSelectionStrategy,
} from '../generated/api';


/**
  * @typedef {import("../generated/api").CartInput} RunInput
  * @typedef {import("../generated/api").CartLinesDiscountsGenerateRunResult} CartLinesDiscountsGenerateRunResult
  */

/**
  * @param {RunInput} input
  * @returns {CartLinesDiscountsGenerateRunResult}
  */

export function cartLinesDiscountsGenerateRun(input) {
  const rawRedeemedPoints = input.cart.attribute?.value?.trim();
  const redeemedPoints = rawRedeemedPoints ? Number.parseInt(rawRedeemedPoints, 10) : NaN;

  if (!Number.isInteger(redeemedPoints) || redeemedPoints <= 0) {
    return {operations: []};
  }

  const hasOrderDiscountClass = input.discount.discountClasses.includes(
    DiscountClass.Order,
  );
  if (!hasOrderDiscountClass) {
    return {operations: []};
  }

  const settingsJson = input.discount.metafield?.jsonValue;
  const rewardValuePerPointRaw =
    settingsJson && typeof settingsJson === 'object'
      ? settingsJson.rewardValuePerPoint
      : undefined;
  const rewardValuePerPoint = Number(rewardValuePerPointRaw);

  if (!Number.isFinite(rewardValuePerPoint) || rewardValuePerPoint <= 0) {
    return {operations: []};
  }

  const discountAmount = redeemedPoints * rewardValuePerPoint;
  if (discountAmount <= 0) {
    return {operations: []};
  }

  return {
    operations: [
      {
        orderDiscountsAdd: {
          selectionStrategy: OrderDiscountSelectionStrategy.First,
          candidates: [
            {
              message: 'Loyalty points redemption',
              targets: [
                {
                  orderSubtotal: {
                    excludedCartLineIds: [],
                  },
                },
              ],
              value: {
                fixedAmount: {
                  amount: discountAmount,
                },
              },
            },
          ],
        },
      },
    ],
  };
}